### PR TITLE
Widget 0xdead10cc fixes part III: it was the old widgets all along

### PR DESCRIPTION
**Pull request by joewalsh**
**Fixes Phabricator ticket:** 
https://phabricator.wikimedia.org/T263247

### Notes
* Updates old widgets to use new widget fetching method
* Fixes retain cycles in image fetching exposed by image fetching via the `UIImageView` extension in the old widgets

### Test Steps
1. Simulator > Erase all content and settings
2. Restart Xcode
3. Set a breakpoint on [MWKDataStore shared]
4. Run the old widget targets
5. Ensure the breakpoint isn't hit
6. Inspect the memory graph
7. Ensure there are no data stores or `NSManagedObjectContext`s remaining

### CODE REVIEW

File: FeaturedArticleWidget/FeaturedArticleWidget.swift [hunk 1]
```diff
@@ -94,35 +81,65 @@ class FeaturedArticleWidget: ExtensionViewController, NCWidgetProviding {
...
+            let siteURL = dataStore.languageLinkController.appLanguage?.siteURL()
+            moc.perform {
+                guard let featuredContentGroup = moc.newestGroup(of: .featuredArticle, forSiteURL: siteURL),
+                    let articleURL = featuredContentGroup.contentPreview as? URL else {
+                    completionHandler(.noData)
+                    return
+                }
+                let article = moc.fetchArticle(with: articleURL)
+                self.update(with: article, dataStore: dataStore) { result in
+                    completionHandler(result ? .newData : .noData)
```
Comment (tonisevener) [inline FeaturedArticleWidget/FeaturedArticleWidget.swift:98]: I think you need to run `completion` here, not `completionHandler`? I get a bit confused on when each needs to be called.

File: InTheNewsNotification/WMFInTheNewsNotificationViewController.swift [hunk 2]
```diff
@@ -134,11 +134,14 @@ class WMFInTheNewsNotificationViewController: ExtensionViewController, UNNotific
...
-            dataStore.savedPageList.addSavedPage(with: articleURL)
-            self.statusView.isHidden = false
-            self.statusLabel.text = WMFLocalizedString("status-saved-for-later", value:"Saved for later", comment: "Indicates to the user that the article has been saved for later")
-            completion(.dismiss)
+            WidgetController.shared.startWidgetUpdateTask(completion) { (dataStore, completion) in
+                DispatchQueue.main.async {
+                    dataStore.savedPageList.addSavedPage(with: articleURL)
+                    self.statusView.isHidden = false
+                    self.statusLabel.text = WMFLocalizedString("status-saved-for-later", value:"Saved for later", comment: "Indicates to the user that the article has been saved for later")
+                    completion(.dismiss)
```
Comment (tonisevener) [inline InTheNewsNotification/WMFInTheNewsNotificationViewController.swift:142]: Some more completion confusion on my end, seems like you should call the inner completion to release the `dataStore` (the one returned next to `dataStore` in line 137), it's unclear if that would happen.

File: TopReadWidget/WMFTodayTopReadWidgetViewController.swift [hunk 3]
```diff
@@ -140,9 +126,19 @@ class WMFTodayTopReadWidgetViewController: ExtensionViewController, NCWidgetProv
...
         chevronImageView.tintColor = theme.colors.secondaryText
     }
-    
+
     @objc func updateView() {
+        let completion = { (result: Bool) in
+            DDLogDebug("Widget did finish")
+        }
+        WidgetController.shared.startWidgetUpdateTask(completion) { (dataStore, completion) in
+            self.updateViewAsync(with: dataStore, completion: completion)
```
Comment (tonisevener) [inline TopReadWidget/WMFTodayTopReadWidgetViewController.swift:135]: Similar completion confusion here.

File: FeaturedArticleWidget/FeaturedArticleWidget.swift [hunk 4]
```diff
@@ -166,17 +183,24 @@ class FeaturedArticleWidget: ExtensionViewController, NCWidgetProviding {
...
     
     @objc func saveButtonPressed() {
-        guard let article = self.article, let articleKey = article.key else {
+        guard let articleKey = articleURL?.wmf_databaseKey else {
             return
         }
-        let isSaved = dataStore?.savedPageList.toggleSavedPage(forKey: articleKey) ?? false
-        expandedArticleView.saveButton.saveButtonState = isSaved ? .longSaved : .longSave
+        WidgetController.shared.startWidgetUpdateTask { (done: Bool) in
+
```
Comment (tonisevener) [inline FeaturedArticleWidget/FeaturedArticleWidget.swift:190]: Might be good to have a `DDLogDebug` here similar to how you did `DDLogDebug("Widget did finish")` for the empty `userCompletion` block below.

File: TopReadWidget/WMFTodayTopReadWidgetViewController.swift [hunk 5]
```diff
@@ -5,14 +5,9 @@ import WMF
 class WMFTodayTopReadWidgetViewController: ExtensionViewController, NCWidgetProviding {
     
     // Model
-    var siteURL: URL!
     var groupURL: URL?
+    var siteURL: URL?
```
Comment (tonisevener) [inline TopReadWidget/WMFTodayTopReadWidgetViewController.swift:9]: To simplify, is there a benefit to having this in a property? Every context I've found also has a `dataSource`, maybe you could pull it from there each time.

File: FeaturedArticleWidget/FeaturedArticleWidget.swift [hunk 6]
```diff
@@ -94,35 +81,65 @@ class FeaturedArticleWidget: ExtensionViewController, NCWidgetProviding {
...
+    }
+    
+    func update(with article: WMFArticle?, dataStore: MWKDataStore, completion: ((Bool) -> Void)?  = nil) {
+        collapsedArticleView.imageView.wmf_imageController = dataStore.cacheController
+        expandedArticleView.imageView.wmf_imageController = dataStore.cacheController
+        articleURL = article?.url
+        guard let article = article,
             let articleKey = article.key else {
                 isEmptyViewHidden = false
-                completionHandler(.failed)
```
[the remark this answers is not in the document]
Comment (joewalsh) [reply FeaturedArticleWidget/FeaturedArticleWidget.swift:104]: This is OK - it's telling the system what happened and there's no user difference.

File: FeaturedArticleWidget/FeaturedArticleWidget.swift (hunk 1 above)
Comment (joewalsh) [reply FeaturedArticleWidget/FeaturedArticleWidget.swift]: That's right, good catch!

File: InTheNewsNotification/WMFInTheNewsNotificationViewController.swift (hunk 2 above)
Comment (joewalsh) [reply InTheNewsNotification/WMFInTheNewsNotificationViewController.swift]: The variable name is shaded here so they're both `completion` - I'll rename it for clarity

**Review by tonisevener — APPROVED** [review]

**Pull request merged**